### ISSUE TITLE
Plugin: Version and requirements

### ISSUE THREAD

**Issue by olleharstedt** [SamMousa/ls3-question-objects-spec#9]
We need member variables to tell the plugin system which version the plugin is. 

Also, requirements. E.g., PHP version, database brand, LimeSurvey release, other installed plugins. Thoughts?

**Comment by SamMousa** [comment]
I don't think we should "manage" those in LS. The only thing i'd like to manage is some kind of API version which is required by the plugin.
The theory is that 1 LS version can then provide multiple api versions for backwards compatibility

**Comment by olleharstedt** [comment]
Can you expand on API with regard to LS release version requirement?

**Comment by SamMousa** [comment]
Not sure what you mean?

**Comment by olleharstedt** [comment]
Can you explain more? ;)

**Comment by Shnoulle** [comment]
At minimum : We should SHOW version number if exist of plugin in plugin management. It's not really difficult.

**Comment by olleharstedt** [comment]
Yes, @Shnoulle, version number and status. "N/A" if no version is given.

**Comment by SamMousa** [comment]
Agreed!

**Comment by olleharstedt** [comment]
So put it in class or config.json? Any logic regarding which should be put where?

**Comment by SamMousa** [comment]
Static information that is needed before code execution should be in json.

**Comment by SamMousa** [comment]
At some point I'd like to support code signing as well to increase security.

**Comment by Shnoulle** [comment]
null/empty is always better if no version given, no ? (i really dislike N/A)

**Comment by SamMousa** [comment]
IMO require it with same spec as composer does.

**Comment by olleharstedt** [comment]
@Shnoulle If it's just empty, the user won't know why it's empty. That's why I'd prefer `N/A` or an icon.

Everybody OK if I add support for plugin version in config.json, then?

```json
{
  'version': '1.2.3',
  'status': 'stable'
}
```

**Comment by SamMousa** [comment]
Dont add status, just use semantic versioning

**Comment by olleharstedt** [comment]
The users won't know about that. How will be communicate status to them?

**Comment by olleharstedt** [comment]
Also sem ver cannot be enforced, but `status` can be forced to have to have one of five different values.

**Comment by SamMousa** [comment]
Semver can be enforced. There are plenty of libraries for it.

**Comment by SamMousa** [comment]
When I parse a semantic version string I can show the user if it's stable.

**Comment by olleharstedt** [comment]
I'm not convinced.

**Comment by olleharstedt** [comment]
Seems like over-engineering, IMHO.

**Comment by olleharstedt** [comment]
I mean,  an LS plugin is not such a serious business.

**Comment by olleharstedt** [comment]
Also, raising the level of required knowledge for no particulary good reason.

**Comment by olleharstedt** [comment]
(But we could at least enforce numbers and `.` as only possible characters in version string.)

**Comment by c-schmitz** [comment]
Use Semantic versioning

**Comment by olleharstedt** [comment]
Well, then we need to include a library, as Sam said. Here's the top google result: https://github.com/naneau/semver

**Comment by SamMousa** [comment]
https://github.com/composer/semver

Top result on  packagist.

**Comment by c-schmitz** [comment]
What do we need a library for?

**Comment by olleharstedt** [comment]
Did not know. Looks fine to me. Most important for me is the possibility to add `alpha` or `beta`.

But we also need a function to verify that the version string is indeed a "PHP-standard" version, so people doesn't by mistake provide version "1" (which is < "1.0").

**Comment by SamMousa** [comment]
We don't need a library but we should use one instead of reinventing the
wheel.

**Comment by olleharstedt** [comment]
Using built-in PHP functions is not reinventing the wheel.

**Comment by SamMousa** [comment]
To parse a semantic version string, there is no function for that in php
afaik?

**Comment by olleharstedt** [comment]
No, not semantic. But "PHP-standard" might be good enough ("1.0.0-dev" etc).

**Comment by LouisGac** [comment]
Is this question really related to LimeSurvey?
No. It's a framework question, which became a PHP question. 
The solution is to use composer : 
http://www.yiiframework.com/wiki/392/creating-yii-applications-with-composer/

**Comment by SamMousa** [comment]
LS3 already uses composer and that is not a solution or even related to this issue @LouisGac.

@olleharstedt I think PHP-standard would also work fine, the thing is though, that most PHP developers use composer a lot nowadays and will be more familiar with the composer version standard.

Also `version_compare` in PHP only does comparing and there are no functions for validation or deriving information like "is this a stable version?".

**Comment by olleharstedt** [comment]
Stable would be the absence of `dev`, `alpha` or `beta`, I guess. Or `>= 1.0.0`.

**Comment by SamMousa** [comment]
So then you're going to parse it anyway :p
Is there a specification for PHP version strings? I've googled but can't seem to find it.
If there is not I think semver is a standard we should adhere. The `composer/semver` package is small and it provides things php does not.

**Comment by olleharstedt** [comment]
Yeah, well, a simple parse for the "status" column in plugin manager.

**Comment by SamMousa** [comment]
LS3 uses composer and composer requires `composer/semver`:
https://github.com/composer/composer/blob/1.0/composer.json

So given that we have the library available I think it makes sense to use it.
https://github.com/composer/semver/blob/master/src/VersionParser.php

I don't have any other arguments to add, if you don't either I'm fine with letting @c-schmitz Make a final decision.

**Comment by olleharstedt** [comment]
Been working on this tonight, and I think I will use `config.php`, as Yii does it (no reason to introduce another configuration file type).

**Comment by olleharstedt** [comment]
Think we might need a library to parse version strings like `1.2.3-beta.2` or `2.3.0-rc.1`.

**Comment by SamMousa** [comment]
There is already a json format in use; also there are definitely arguments
for non executable config files.

Why not use what is already in place?

**Comment by SamMousa** [comment]
The main argument for non-executable config is security. The json file could include a signature that verifies the codes, this is not the case for a php file.

In LS3 we use several config files:
1. Yii php files.
2. Composer json file
3. PHPUnit XML file
4. Codeception YAML file.

All of these formats with the exception of Yii's is not executable. In my opinion we should not spend time changing the current json format without clear arguments.

Also arguing we "already" have XML for template config is not valid, when the XML format was introduced we already had a json format for plugins.
Note that the file extension is not that important to me (I prefer json because that has already been implemented), but what is important to me is that a configuration file for 3rd party code should not be executable.

**Comment by SamMousa** [comment]
> Think we might need a library to parse version strings like 1.2.3-beta.2 or 2.3.0-rc.1. Either that or limit the version format to use a subset of semver, which might be enough for plugins.

Yes, composer includes this by default, so we have the library available already.

**Comment by olleharstedt** [comment]
You're assuming your development branch will be in production some day. There's no such decision, so saying "we already have this or that in LS3" makes no sense.

**Comment by SamMousa** [comment]
@c-schmitz ? I thought the only question was whether to merge 2.5 into develop or the other way around, or port features from 2.5 into develop or the other way around?
I hope you're not seriously considering throwing all that work away?

**Comment by olleharstedt** [comment]
Stackoverflow about security of PHP config files: http://stackoverflow.com/a/17666811/2138090

**Comment by SamMousa** [comment]
1. We don't want php parse errors since we can't catch them until php 7.
2. That is why I the parsed json is cached in a php file ;-)
3. The configuration file of a plugin should never contain any private information, it is not a file for storing plugin settings, it is a file documenting how LS core should interact with / load the plugin.

**Comment by olleharstedt** [comment]
2. ?

**Comment by SamMousa** [comment]
The json files from all plugins are parsed and when they are valid they are stored in the configuration directory in LS, allowing for speedy parsing of the configuration information without the risk of including faulty PHP code.

**Comment by c-schmitz** [comment]
@SamMousa I am not considering anything at all - there is no decision on how we will integrate LS2.5 and LS3 codebase yet - that's what @olleharstedt meant.
